stats/stats.go

```diff
@@ -160,11 +161,106 @@ func (t ValueType) String() string {
 	}
 }
 
+// SampleTags is an immutable string[string] map for tags. Once a tag
+// set is created, direct modification is prohibited. It has
+// copy-on-write semantics and uses pointers for faster comparison
+// between maps, since the same tag set is often used for multiple samples.
+// All methods should not panic, even if they are called on a nil pointer.
+type SampleTags struct {
+	tags map[string]string
+	json []byte
+}
+
+// Get returns an empty string and false if the the requested key is not
+// present or its value and true if it is.
+func (st *SampleTags) Get(key string) (string, bool) {
+	if st == nil {
+		return "", false
+	}
+	val, ok := st.tags[key]
+	return val, ok
+}
+
+// IsEqual trues to compare two tag sets with maximum efficiency.
+func (st *SampleTags) IsEqual(other *SampleTags) bool {
+	if st == other {
+		return true
+	}
+	if st == nil || other == nil || len(st.tags) != len(other.tags) {
+		return false
+	}
+	for k, v := range st.tags {
+		if otherv, ok := other.tags[k]; !ok || v != otherv {
+			return false
+		}
+	}
+	return true
+}
+
+// MarshalJSON serializes SampleTags to a JSON string and caches
+// the result. It is not thread safe in the sense that the Go race
+// detector will complain if it's used concurrently, but no data
+// should be corrupted.
+func (st *SampleTags) MarshalJSON() ([]byte, error) {
+	if st == nil {
+		return []byte("null"), nil
+	}
+	if st.json != nil {
+		return st.json, nil
+	}
+	res, err := json.Marshal(st.tags)
+	if err != nil {
+		return res, err
+	}
+	st.json = res
+	return res, nil
+}
+
+// UnmarshalJSON deserializes SampleTags from a JSON string.
+func (st *SampleTags) UnmarshalJSON(data []byte) error {
+	if st == nil {
+		*st = SampleTags{}
+	}
+	return json.Unmarshal(data, &st.tags)
+}
+
+// CloneTags copies and sample tags and underlying tag set returns
+// the result. If the receiver is nil, it returns an empty non-nil map.
```

Comment:
Nit pick, but I think the phrasing in this comment can be tweaked to say something like "CloneTags copies the underlying tag set of a sample tags receiver and returns it. If the receiver is nil, it returns an empty non-nil map." perhaps?

Reply:
Totally agree, will fix in a minute